### DISCUSSION TITLE
Linking against python

### DISCUSSION THREAD

**Discussion by oli747** [jralls/gtk-osx-build#89] [Q&A] [answered]
Hi all,

it has been a while ago since the last time I created a Bluefish package for MacOS. I now have a virtual Big Sur image so I can create a Bluefish package again. However, I run into a linking problem:

ld: cannot link directly with dylib/framework, your binary is not an allowed client of /Applications/Xcode.app/Contents/Developer/blah/blah

How do I work around this? If I cannot link to the system python, should I include a complete python stack in the Bluefish package? (only one bluefish plugin needs libpython)

**Comment by jralls** [comment, marked as answer]
Yes, if you need to use Python in your app you absolutely must build it and bundle that. It's pretty painless, just add `pygments` to your modules list. You need `pygments` instead of `python3` because once you include Python in the build everything that uses python uses *that* python and `gtk-docs` needs `pygments` for its code-highlighting feature.

Figuring out what to bundle is a little more challenging because you need to figure out what the plugin uses. It's easy enough to include `libpython.dylib` and the contents of `$PREFIX/lib/python3.11/`, but do you also need the python interpreter? There's also the problem of linkage: The bundler is hardwired to use `@executable_path/../Resources/lib/`. If the plugin is `dlopen`ed from the Bluefish executable then it's not a problem, but if it's forked and execed then either it needs to be placed so that the linker can find it with that path or you need to make some adjustments in gtk-mac-bundler to get it the right path. ld(1) has detailed info on linkage paths; you'll need to understand that to figure out what to do.

**Reply by oli747** [reply]
ok, this did include a different problem. I use python-config to retrieve the compiler flags (--libs and --cflags). This does contain the include locations for the headers, but not for the libraries. So during linking I get an error: library not found for -lpython3.11

I can set DYLD_LIBRARY_PATH  to ~/gtk/inst/lib but then pkg-config fails to run with a library load failure.. 

So what is the solution to get the correct library path to include and link to pygments?

**Reply by jralls** [reply]
Since somewhere around 3.9 you have to add an `--embed` option to `python-config` to get `-lpython3.x` included in the `--libs` output; They added `python-3.x-embed.pc` too so you have to say `pkg-config --libs --cflags python-3.x-embed`.

**Comment by oli747** [comment]
Thanks, that did the trick. Compiled latest Bluefish 2.2.14 release succesfully. Had to remove a reference to #include <gtk-mac-menu.h> but without that header it still compiled fine. Core program is running, so I'll go test if everything works.

**Reply by jralls** [reply]
> Had to remove a reference to #include <gtk-mac-menu.h> 

Does it also have an `#include <gtkosxapplication.h>`? `gtk-mac-menu.h` was the old Carbon-based code for putting the app's menus in the menu bar and `gtkosxapplication` is the not as old (about 10 years...) Cocoa-based code that replaced it.

**Reply by oli747** [reply]
yes it does

**Comment by oli747** [comment]
> Since somewhere around 3.9 you have to add an `--embed` option to `python-config` to get `-lpython3.x` included in the `--libs` output; They added `python-3.x-embed.pc` too so you have to say `pkg-config --libs --cflags python-3.x-embed`.

but that still does not include the -L~/gtk/inst/lib required to locate that library.. But I found a workaround.

**Reply by jralls** [reply]
True, and there doesn't seem to be a flag to extract it directly. `pkg-config --libs python-3.11-embed` does. OTOH $PREFIX is defined in the build so you can always do something to add `LDFLAGS="$LDFLAGS -L$PREFIX/lib"` to the build environment.

BTW, you need to use `python3-config` to be sure of getting the one built in $PREFIX: Python doesn't make the symlinks so if you don't make them by hand `python-config` will be the one in the virtualenv running jhbuild instead of the built one.
